### PR TITLE
Issue implemented

### PR DESCRIPTION
**Pull request by dipasqualew**
Entire-Checkpoint: d4b842ca1e5f

### CODE REVIEW

File: apps/cli/src/lib/observability.ts [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+ * Get output directory for logs/traces
+ */
+function getOutputDir(branch: string, subdir: string, baseDir = "/tmp/vibx"): string {
+  return path.join(baseDir, branch, subdir);
+}
+
+/**
+ * Create a Winston logger configured for JSONL output
+ */
+export function createLogger(config: ObservabilityConfig): Logger {
```
Comment (dipasqualew) [inline apps/cli/src/lib/observability.ts:125]: :information_source: **Add Winston JSONL logging with output to /tmp/vibx/$branch/logs/vibx-{timestamp}.jsonl**
Status: :white_check_mark: fulfilled

createLogger() creates a Winston logger with format.json() (JSONL) output. The log directory is constructed as path.join(baseDir, branch, 'logs') defaulting to /tmp/vibx, and the file is named vibx-{timestamp}.jsonl. This exactly matches the requirement.

File: apps/cli/src/lib/observability.ts [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+
+const ExportResultCode = {
+  SUCCESS: 0,
+  FAILED: 1,
+};
+
+/**
+ * Custom file-based span exporter that writes JSONL to disk
+ */
+class FileSpanExporter implements SpanExporter {
```
Comment (dipasqualew) [inline apps/cli/src/lib/observability.ts]: :information_source: **Add OpenTelemetry tracing with a custom FileSpanExporter writing JSONL to /tmp/vibx/$branch/traces/trace-{timestamp}.jsonl**
Status: :white_check_mark: fulfilled

FileSpanExporter writes each span as a JSON line to a file named trace-{timestamp}.jsonl inside the traces subdirectory. initTracing() wires it up via SimpleSpanProcessor. The output path matches the requirement.

File: apps/cli/src/lib/observability.ts [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+    return "unknown";
+  }
+
+  return sanitizeBranchName(stdout.trim());
+}
+
+/**
+ * Sanitize branch name for filesystem safety
+ */
+export function sanitizeBranchName(branch: string): string {
```
Comment (dipasqualew) [inline apps/cli/src/lib/observability.ts:111]: :information_source: **Branch name must be sanitized (replace /\:*?"<>| with -) for filesystem safety**
Status: :white_check_mark: fulfilled

sanitizeBranchName() uses the regex /[/\\:*?"<>|]/g to replace all specified characters with hyphens. The test file validates all listed characters plus edge cases.

File: apps/cli/src/lib/observability.ts (hunk 2 above)
Comment (dipasqualew) [inline apps/cli/src/lib/observability.ts]: :warning: **Create src/lib/observability.ts exporting getCurrentBranch(), createLogger(), initTracing(), shutdownObservability(), and FileSpanExporter**
Status: :yellow_circle: partial

getCurrentBranch, createLogger, initTracing, and shutdownObservability are all exported. However, FileSpanExporter is declared as a private class (no export keyword) at line 34. The requirement explicitly lists FileSpanExporter as one of the required exports, but the class is not exported, making it inaccessible to consumers and untestable directly. The test file does not test FileSpanExporter directly either, relying instead on testing withSpan via a mock tracer.

File: apps/cli/src/lib/context.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
+import {
```
Comment (dipasqualew) [inline apps/cli/src/lib/context.ts:1]: :information_source: **Create src/lib/context.ts with an AppContext interface and createAppContext() factory for dependency injection**
Status: :white_check_mark: fulfilled

AppContext interface exports logger (Logger) and tracer (Tracer) and branch (string). createAppContext() is an async factory that calls getCurrentBranch(), createLogger(), and initTracing(), then returns a composed context. This matches the dependency injection pattern required.

File: apps/cli/src/index.ts [hunk 5]
```diff
@@ -10,39 +12,85 @@ import {
...
+    ctx?: AppContext;
+  }
+}
+
+async function main() {
+  // Initialize observability
+  const ctx = await createAppContext();
+
+  ctx.logger.info("CLI started", { args: process.argv.slice(2) });
+
```
Comment (dipasqualew) [inline apps/cli/src/index.ts:27]: :information_source: **Update src/index.ts to initialize observability and inject context via yargs middleware, with a shutdown hook for graceful trace flushing**
Status: :white_check_mark: fulfilled

main() calls createAppContext(), injects ctx into yargs via .middleware(), and registers SIGINT/SIGTERM handlers plus a finally block that calls shutdownObservability(). Graceful shutdown is properly implemented.

File: apps/cli/src/commands/review.ts [hunk 6]
```diff
@@ -37,71 +39,86 @@ interface ReviewOptions {
...
+  skillName: string
+): Promise<Record<string, unknown>> {
+  return withSpan(ctx.tracer, "runReviewer", async () => {
+    const env = Object.fromEntries(
+      Object.entries(process.env).filter(([k]) => k !== "CLAUDECODE")
+    );
+
+    const cmd = ["claude", "-p", `/${skillName} ci`, "--output-format", "json"];
+    ctx.logger.info("Running reviewer command", { command: cmd.join(" ") });
+    console.error(`Running command: ${cmd.join(" ")}`);
```
Comment (dipasqualew) [inline apps/cli/src/commands/review.ts:53]: :grey_question: **Update all four command handlers (review.ts, intent.ts, get-intents.ts, get-checkpoint-folders.ts) to replace console.error with logger and add spans around key operations**
Status: :yellow_circle: partial

Logger calls have been added throughout all four command handlers and key operations are wrapped with withSpan/tracer.startActiveSpan. However, multiple console.error calls remain in review.ts (lines 53, 70-72, 76, 86-87, 94, 118-119, 173, 236) alongside their logger equivalents rather than being replaced. Similarly in get-intents.ts (line 12) and intent.ts (line 175, 183), console.error calls are kept in parallel with new logger calls. The requirement says to 'replace' console.error with the logger, not supplement it. That said, some of these may be intentional for stderr CLI output visibility.

File: apps/cli/src/lib/github.ts [hunk 7]
```diff
@@ -32,77 +33,85 @@ async function runCommand(cmd: string[]): Promise<ProcessResult> {
...
-    if (eventPath) {
-      try {
-        const file = Bun.file(eventPath);
-        const event = await file.json();
-        const prNumber = event?.pull_request?.number;
-        if (typeof prNumber === "number") {
-          return prNumber;
+  const tracer = getTracer();
+
+  return withSpan(tracer, "getCurrentPR", async () => {
```
Comment (dipasqualew) [inline apps/cli/src/lib/github.ts:38]: :information_source: **Update src/lib/github.ts to add spans around getCurrentPR, getDiffLines, submitReview, and replace console.error with logger**
Status: :white_check_mark: fulfilled

getCurrentPR, getDiffLines, getCommitSha, and submitReview are all wrapped with withSpan(). All previous console.error calls in github.ts have been replaced with logger?.error() calls (using the optional logger parameter pattern). No raw console.error calls remain in github.ts.

File: apps/cli/tests/observability.test.ts [hunk 8]
```diff
@@ -0,0 +1,235 @@
+import { describe, it, expect, beforeEach, afterEach } from "vitest";
```
Comment (dipasqualew) [inline apps/cli/tests/observability.test.ts:1]: :information_source: **Add unit tests for observability utilities**
Status: :white_check_mark: fulfilled

26 tests cover sanitizeBranchName (14 cases for all specified special characters and edge cases), createLogger (3 tests for interface, directory creation, and branch sanitization), and withSpan (9 tests for return value, OK status, span ending, attributes, error status, exception recording, non-Error throws, and span naming). Coverage is thorough for the tested utilities.

File: apps/cli/package.json [hunk 9]
```diff
@@ -13,6 +13,12 @@
     "test:run": "vitest run"
   },
   "dependencies": {
+    "@opentelemetry/api": "^1.9.0",
+    "@opentelemetry/resources": "^1.30.0",
```
Comment (dipasqualew) [inline apps/cli/package.json:17]: :grey_question: **@opentelemetry/sdk-node declared in package.json but not used at runtime (drift from plan)**
Status: :x: violation

package.json declares @opentelemetry/sdk-node as a production dependency at line 17, but the implementation does not import or use it anywhere in the source code. The implementing engineer intentionally dropped it in favor of @opentelemetry/sdk-trace-base, but forgot to remove it from package.json. This adds unnecessary bloat to the production dependency list and is misleading. It should be removed.

File: apps/cli/src/lib/observability.ts (hunk 2 above)
Comment (dipasqualew) [inline apps/cli/src/lib/observability.ts]: :warning: **FileSpanExporter not exported from observability.ts**
Status: :x: violation

The FileSpanExporter class is declared private (no export modifier). The requirement explicitly states it should be exported. This means consumers cannot import it directly for testing or extension, and no direct unit tests exist for it. The test suite tests withSpan with a mock tracer but never exercises the actual file writing behavior of FileSpanExporter. Consider adding `export` to the class declaration and adding integration tests that verify actual file output.

**Review by dipasqualew — COMMENTED** [review]
# Vibereq Review

## code-review
**Overall Status:** :warning: WARN

Implementation fulfills all core requirements with one notable gap: FileSpanExporter is not exported from observability.ts as required, and @opentelemetry/sdk-node is declared in package.json but not used at runtime.

### General Findings

### :information_source: Preserve intentional console.log calls used for CLI output to users
**Status:** :white_check_mark: fulfilled | **Severity:** info

All console.log calls for user-facing output (e.g., folder paths in get-checkpoint-folders.ts, review status in review.ts, PR URLs, dry-run previews) are preserved. New logger calls are added alongside them for structured logging, not as replacements.

### :information_source: Build must pass (bun run build)
**Status:** :white_check_mark: fulfilled | **Severity:** info

Build succeeds: 313 modules bundled and compiled to dist/vibx in ~189ms with no errors.

### :information_source: All tests must pass (bun run test)
**Status:** :white_check_mark: fulfilled | **Severity:** info

All 43 tests pass across 3 test files: git.test.ts (6), transcript.test.ts (11), observability.test.ts (26). This matches the implementing engineer's report.

File: apps/cli/src/lib/observability.ts [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+
+/**
+ * Create a Winston logger configured for JSONL output
+ */
+export function createLogger(config: ObservabilityConfig): Logger {
+  const logDir = getOutputDir(config.branch, "logs", config.baseDir);
+  fs.mkdirSync(logDir, { recursive: true });
+
+  const timestamp = new Date().toISOString().replace(/[:.]/g, "-");
+  const logFile = path.join(logDir, `vibx-${timestamp}.jsonl`);
```
Comment (dipasqualew) [inline apps/cli/src/lib/observability.ts:130]: :information_source: **Winston JSONL logging writes structured output to /tmp/vibx/{branch}/logs/vibx-{timestamp}.jsonl**
Status: :white_check_mark: fulfilled

createLogger() in observability.ts creates a JSONL log file at the correct path using getOutputDir(branch, 'logs', baseDir). Format is json() via Winston. Branch sanitization is applied before path construction.

**Review by dipasqualew — COMMENTED** [review]
# Vibereq Review

## code-review
**Overall Status:** :white_check_mark: PASS

Observability implementation fulfills all core requirements with minor residual console.error calls in command handlers that do not violate the stated intent

### General Findings

### :grey_question: console.log is preserved for intentional CLI output; console.error/debug calls in command and lib files are replaced with logger calls
**Status:** :yellow_circle: partial | **Severity:** minor

All intentional console.log calls for CLI output are correctly preserved. Logger calls have been added alongside all errors/warnings. However, in get-checkpoint-folders.ts line 33, get-intents.ts lines 12/25/31/60/86/111/118, intent.ts lines 175/183, and review.ts lines 53/70/71/72/76/86/87/94/118/119/173/236, console.error calls are retained in addition to logger calls rather than being replaced. The requirement states console.error calls should be replaced, but the implementation keeps both. This is a minor issue since the logger calls are present; the console.error calls create duplicate output on stderr.

### :information_source: Build (bun run build) succeeds with no TypeScript errors
**Status:** :white_check_mark: fulfilled | **Severity:** info

Build confirmed passing: 313 modules bundled, compiled successfully to dist/vibx in ~118ms.

### :information_source: All existing and new tests pass (bun run test)
**Status:** :white_check_mark: fulfilled | **Severity:** info

43 tests pass across 3 test files: git.test.ts (6), transcript.test.ts (11), observability.test.ts (26). New observability tests cover sanitizeBranchName, createLogger directory creation, and withSpan behavior including success, error, attribute passing, and span lifecycle.

File: apps/cli/src/lib/observability.ts [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+  });
+}
+
+let provider: BasicTracerProvider | null = null;
+
+/**
+ * Initialize OpenTelemetry tracing
+ */
+export function initTracing(config: ObservabilityConfig): Tracer {
+  const traceDir = getOutputDir(config.branch, "traces", config.baseDir);
```
Comment (dipasqualew) [inline apps/cli/src/lib/observability.ts:158]: :information_source: **OpenTelemetry traces write JSONL span data to /tmp/vibx/{branch}/traces/trace-{timestamp}.jsonl**
Status: :white_check_mark: fulfilled

initTracing() creates a FileSpanExporter targeting the correct traces subdirectory. FileSpanExporter writes JSONL via a WriteStream. No OTLP collector dependency.

File: apps/cli/src/lib/observability.ts [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+    return "unknown";
+  }
+
+  return sanitizeBranchName(stdout.trim());
+}
+
+/**
+ * Sanitize branch name for filesystem safety
+ */
+export function sanitizeBranchName(branch: string): string {
```
Comment (dipasqualew) [inline apps/cli/src/lib/observability.ts:111]: :information_source: **Git branch name is sanitized (replace /\:*?"<>| with -) before use as a filesystem path segment**
Status: :white_check_mark: fulfilled

sanitizeBranchName() applies the regex /[/\\:*?"<>|]/g to replace all listed characters with hyphens. getCurrentBranch() calls this before returning. Tests in observability.test.ts validate all expected character substitutions.

File: apps/cli/src/lib/observability.ts [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+
+const ExportResultCode = {
+  SUCCESS: 0,
+  FAILED: 1,
+};
+
+/**
+ * Custom file-based span exporter that writes JSONL to disk
+ */
+export class FileSpanExporter implements SpanExporter {
```
Comment (dipasqualew) [inline apps/cli/src/lib/observability.ts:34]: :information_source: **A custom FileSpanExporter class handles JSONL span output (no OTLP collector required)**
Status: :white_check_mark: fulfilled

FileSpanExporter implements SpanExporter from @opentelemetry/sdk-trace-base. Uses only @opentelemetry/sdk-trace-base, not the heavier @opentelemetry/sdk-node package (intentional deviation noted in drift section).

File: apps/cli/src/lib/context.ts [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+  createLogger,
+  initTracing,
+  getCurrentBranch,
+  Logger,
+  Tracer,
+} from "./observability.js";
+
+export interface AppContext {
```
Comment (dipasqualew) [inline apps/cli/src/lib/context.ts:9]: :information_source: **AppContext interface (logger, tracer, branch) is created as a dependency injection container**
Status: :white_check_mark: fulfilled

AppContext interface defines logger (Winston Logger), tracer (OTel Tracer), and branch (string). createAppContext() factory assembles all three and is the sole entry point.

File: apps/cli/src/index.ts [hunk 15]
```diff
@@ -10,39 +12,85 @@ import {
...
+  });
+  process.on("SIGTERM", async () => {
+    await shutdown();
+    process.exit(143);
+  });
+
+  try {
+    await yargs(hideBin(process.argv))
+      .scriptName("vibx")
+      .middleware((argv) => {
```
Comment (dipasqualew) [inline apps/cli/src/index.ts:49]: :information_source: **Context is injected into command handlers via a yargs middleware**
Status: :white_check_mark: fulfilled

yargs .middleware() sets argv.ctx = ctx before each command runs. All four command handlers (get-checkpoint-folders, get-intents, intent, review) receive ctx via argv.ctx!.

File: apps/cli/src/commands/get-intents.ts [hunk 16]
```diff
@@ -1,27 +1,33 @@
+import type { AppContext } from "../lib/context.js";
```
Comment (dipasqualew) [inline apps/cli/src/commands/get-intents.ts:1]: :grey_question: **Spans are added around claude subprocess calls (review, intent, get-intents) and GitHub API calls (getCurrentPR, getDiffLines, submitReview)**
Status: :yellow_circle: partial

review.ts wraps runReviewer, getDiffLines, submitReview, and the top-level review flow with withSpan. intent.ts wraps processIntents and the runClaude call with withSpan. github.ts wraps getCurrentPR, getDiffLines, getCommitSha, submitReview, and gh.api.createReview with withSpan. However, get-intents.ts has no span instrumentation around its key operations (getCheckpointFolders, generateIntents), which means the 'get-intents' command flow has no trace coverage despite being listed in the requirement.

File: apps/cli/src/index.ts [hunk 17]
```diff
@@ -10,39 +12,85 @@ import {
...
+
+  ctx.logger.info("CLI started", { args: process.argv.slice(2) });
+
+  // Register shutdown hooks for signals only
+  let shuttingDown = false;
+  const shutdown = async () => {
+    if (shuttingDown) return;
+    shuttingDown = true;
+    ctx.logger.info("CLI shutting down");
+    await shutdownObservability();
```
Comment (dipasqualew) [inline apps/cli/src/index.ts:34]: :information_source: **shutdownObservability() ensures graceful trace flushing on process exit**
Status: :white_check_mark: fulfilled

shutdownObservability() calls provider.shutdown() which flushes the SimpleSpanProcessor. It is called in the finally block of main() and also on SIGINT/SIGTERM signals. The shuttingDown guard prevents double-shutdown.

File: apps/cli/src/lib/observability.ts [hunk 18]
```diff
@@ -0,0 +1,227 @@
...
+  if (provider) {
+    await provider.shutdown();
+    provider = null;
+  }
+}
+
+/**
+ * Create a span and execute a function within it
+ */
+export async function withSpan<T>(
```
Comment (dipasqualew) [inline apps/cli/src/lib/observability.ts:196]: :information_source: **withSpan helper (scope creep - not in original plan)**
Status: :white_check_mark: fulfilled

withSpan() was added as an undocumented additive deviation from the original plan. It is a clean, well-typed utility that reduces boilerplate for async span wrapping. It handles status setting, exception recording, and span.end() in finally. This is a beneficial addition that does not conflict with any stated requirement.

File: apps/cli/src/lib/github.ts [hunk 19]
```diff
@@ -32,77 +33,85 @@ async function runCommand(cmd: string[]): Promise<ProcessResult> {
...
-  if (process.env.GITHUB_EVENT_NAME === "pull_request") {
-    const eventPath = process.env.GITHUB_EVENT_PATH;
-    if (eventPath) {
-      try {
-        const file = Bun.file(eventPath);
-        const event = await file.json();
-        const prNumber = event?.pull_request?.number;
-        if (typeof prNumber === "number") {
-          return prNumber;
+  const tracer = getTracer();
```
Comment (dipasqualew) [inline apps/cli/src/lib/github.ts:36]: :grey_question: **Tracer injection consistency - github.ts uses module-level getTracer() instead of AppContext**
Status: :x: violation

github.ts uses getTracer() (a module-level singleton getter) rather than receiving the tracer via dependency injection through AppContext. This is inconsistent with the DI pattern established by AppContext and makes testing harder. The CLAUDE.md guidelines explicitly require dependency injection. submitReview() receives a Logger via optional parameter but not a Tracer, and getCurrentPR()/getDiffLines() obtain their tracer from the global singleton. This should use a passed-in tracer for consistency and testability.
